client/app.jsx

```diff
@@ -0,0 +1,114 @@
+      i18n = require('i18next-client'),
+      io = require('socket.io-client'),
+      _ = require('lodash'),
+      config = require('./config.jsx'),
+      Route = require('react-router-dom').Route,
+      BrowserRouter = require('react-router-dom').BrowserRouter,
+      Switch = require('react-router-dom').Switch,
+      TopBar = require('./topbar.jsx'),
+      Ting = require('./ting.jsx'),
+      Settings = require('./settings.jsx');
+
+const App = React.createClass({
+    _socket: null,
+    getInitialState() {
+        return {
+            username: null,
+            people: [],
+            currentMessageId: null
+        };
+    },
+    componentWillMount() {
+        const URL = window.location.hostname + ':' + config.port;
+        this._socket = io.connect(URL, {secure: config.websocket.secure});
+    },
+    updateUsername(username) {
+        this.setState({username});
+    },
+    updateCurrentMessageId(messageid) {
+        this.setState({currentMessageId: messageid});
+    },
+    onLoginIntention(username, people) {
+        this.setState({username, people});
+    },
+    listenOnLoginResponse(onError, onSuccess, onLogin, username) {
+        this._socket.on('login-response', ({success, people, error}) => {
+            if (!success) {
+                onError(error);
+            }
+            else {
+                onSuccess();
+
+                var peopleList = _.chain(people)
+                    .value();
+
+                onLogin(this.state.username, peopleList);
+            }
+        });
+    },
+    listenOnMessage(onMessage) {
+        this._socket.on('message', (data) => {
+            onMessage(data);
+        });
+    },
+    listenOnPart(onPart, deleteTypingMessage) {
+        this._socket.on('part', (username) => {
+            onPart(username);
+            deleteTypingMessage(username);
+        });
+    },
+    listenOnJoin(onJoin) {
+        this._socket.on('join', (username) => {
+            onJoin(username);
+        });
+    },
+    listenOnStartTypingResponse(onStartTypingResponse) {
+        this._socket.on('start-typing-response', (messageid) => {
+            this.setState({currentMessageId: messageid});
+            onStartTypingResponse(messageid);
+        });
+    },
+    listenOnUpdateTypingMessages(onUpdateTypingMessages) {
+        this._socket.on('update-typing-messages', (messagesTyping) => {
+            onUpdateTypingMessages(messagesTyping);
+        });
+    },
+    render() {
+        return (
+            <BrowserRouter>
+                <Switch>
+                    <Route exact path='/' render={(props) => (
+                        <Ting {...props} username={this.state.username}
+                                         people={this.state.people}
+                                         onLoginIntention={this.onLoginIntention}
+                                         updateUsername={this.updateUsername}
+                                         _socket={this._socket}
```

Comment:
Underscore-prefixed properties are reserved as private. Therefore it is not proper to name the `_socket` property of the `Ting` component as `_socket`. Instead you can name it `_socket`. Component props are inherently non-private.